### PR TITLE
Add CouchDB session authenticator

### PR DESCRIPTION
**Pull request by eiri**
## PR summary

This PR adds `auth` module that implements authenticator for CouchDB session authentication

Fixes: #4 

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?
CouchDB session authentication is not supported.

## What is the new behavior?
CouchDB session authentication is not supported for both env file and programmatic approach

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

New tests can be run in isolation as `go test -v ./auth/...`

### CODE REVIEW

**Review by emlaver — COMMENTED** [review]
First pass of review: I suggest we change all `a` parameters in the method definitions

File: auth/authenticator.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+// AuthenticationType returns the authentication type for this authenticator.
+func (a CouchDbSessionAuthenticator) AuthenticationType() string {
+	return AUTHTYPE_COUCHDB_SESSION
+}
+
+// Validate the authenticator's configuration.
+// Ensures the service url, username and password are valid and not nil.
+func (a CouchDbSessionAuthenticator) Validate() error {
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go:87]: This is against golang's style conventions though, using a one-letter variable for a receiver is extremely common and is generally recommended in [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#variable-names). The reasoning goes like:

> The greater the distance between a name's declaration and its uses,
the longer the name should be. 

Ofc we can ignore go's convention, but I'm not sure how beneficial is it - we can't use `authenticator` as a var name, because this is a name of the package and I don't feel that `auth` makes things that much more readable, given that most of the functions is ~8 lines long and var's descriptive type is right here in the method definition.

[the remark this answers is not in the document]
Comment (emlaver) [reply auth/authenticator.go:87]: Ok interesting, Go seems to be the opposite compared to other language styles that suggest using meaningful variable names.

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go:87]: Yes, golang's naming is rather controversial, influenced by Rob Pike's (one of the creators) philosophy, as far as I understand. There are a couple presentations on the topic around, e.g. [https://talks.golang.org/2014/names.slide#1](https://talks.golang.org/2014/names.slide#1)

Personally this felt rather strange when I was starting with the language, but now it's other way around, for example I find `core`'s style is very unnatural.

That said I still can switch from `a` to `auth` if you think it'll really increase readability and not just a question of naming preference coming from other languages.

[the remark this answers is not in the document]
Comment (emlaver) [reply auth/authenticator.go:87]: No need to change it, let's stick with the naming convention.

[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go:87]: @eiri What do you think about the variable `this`? As I can see this is used also in the core: https://github.com/IBM/go-sdk-core/blob/775201f679d69afb97f3259ababb24d482089410/v4/core/basic_authenticator.go#L76

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go:87]: @vmatyusGitHub I dislike javaisms in go code, I think it's misleading to bring in primitives from other languages, especially from the languages with a different paradigm.

As I said, `core`'s style looks very unnatural to me and I'd prefer not to mimic it, unless it's really necessary for users familiarity.

File: auth/authenticator_test.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			cookie, err := auth.getCookie()
```
Comment (vmatyusGitHub) [inline auth/authenticator_test.go:146]: This is the first fetch from the cache, better to assert this, like from line 154 till 159. After you make this change the second fetch and its checks are not needed anymore.

File: auth/authenticator_test.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+			}
+			wg.Wait()
+			close(done)
+		})
+
+		It("Test authentication refresh", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			cookie, err := auth.getCookie()
```
Comment (vmatyusGitHub) [inline auth/authenticator_test.go:186]: I do not see what is the point of this command. This fetches `cookie` from the cache.
If you want to store the cookie value, I would suggest to do this way:
```suggestion
			oldCookie := &auth.session.cookie
```

File: auth/authenticator_test.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+			Expect(err).To(BeNil())
+			Expect(cookie).ToNot(BeNil())
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+
+			// Move time into the refresh window
+			auth.session.expires = time.Now().Add(30 * time.Minute)
```
Comment (vmatyusGitHub) [inline auth/authenticator_test.go:192]: This test case focus on `refreshTime` no need to manipulate anything with expires, so this line can be removed.

**Review by vmatyusGitHub — CHANGES_REQUESTED** [review]
Very well done! I can see the care and the thoroughness in your work!

However, still find things to do:

How is the URL from the base service is transferred to the authenticator?
You should consider to create also a `CoudantBaseService` with through you can also change the authenticator URL, when the url of the base service is changed.
Please also include `ConfigureService` function.

Also do not forget to transfer the default headers from BaseService.

Regrading testing I also would like to request at least 2 integration test example where you can check the new `CloudantBaseService` works properly with the new `GetAuthenticatorFromEnvironment`.
One of the should create client with `NewCloudantV1UsingExternalConfig` and the second one use `NewCloudantV1`. - These tests will be not merged into the master, but would help me in the review and you in the development. This will help to depict the scenario, how the user should instantiate a client with session authenticator and how can change the details of the service.

File: auth/authenticator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go:87]: Okay, I can accept this point of view. Just was an idea.

**Comment by vmatyusGitHub** [comment]
Here I am listing the already implemented `CloudantBaseService` variations:

* node: https://github.com/IBM/cloudant-node-sdk/blob/master/lib/cloudantBaseService.ts
* python: https://github.com/IBM/cloudant-python-sdk/blob/master/ibmcloudant/couchdb_session_base_service_patch.py
* java: https://github.com/IBM/cloudant-java-sdk/blob/master/modules/common/src/main/java/com/ibm/cloud/cloudant/internal/CloudantBaseService.java

Just to have it here, as an inspiration.

File: auth/authenticator.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	AUTHTYPE_COUCHDB_SESSION = "couchdbSession"
+)
+
+var requestSessionMutex sync.Mutex
+
+// CouchDbSessionAuthenticator uses username and password to obtain
+// CouchDB authentication cookie, and adds the cookie to requests.
+type CouchDbSessionAuthenticator struct {
+	URL, Username, Password string
+	Client                  *http.Client
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go:38]: We had this conversation on Slack about this and agreed between @emlaver and @ricellis that having public `Client` var to allow setting custom `http.Client` is sufficient.

Here is my write-up for possible approaches and cons and pros for this, it includes view on `SetHTTPClient` https://gist.github.com/eiri/f9e6a2a7ed7e0cc6bdf7a8fec76c387b

**Comment by eiri** [comment]
@vmatyusGitHub Well, the thing with `CoudantBaseService` is that unlike java, typescript or python golang doesn't have inheritance, it only have composition, so that'd work [like this](https://gist.github.com/eiri/f9e6a2a7ed7e0cc6bdf7a8fec76c387b#creating-new-struct-cloudantbaseservice-and-use-it-for-cloudantv1-service)

I talked about this last Friday on a meeting as a possibility, we discussed it and I thought we all agreed that in this case adding `CoudantBaseService` it's more hustle than there is a gain, so I'm a bit surprised you are insisting on this now. 

I can change the approach here, but if this is only about setting url and headers on auth request, much simpler way would be to pull them from `http.Request` during `auth.Request`.

Then the injection of auth in `cloudant_v1.go` will be handled like this:

```
    func NewCloudantV1UsingExternalConfig(options *CloudantV1Options) (cloudant *CloudantV1, err error) {
        if options.ServiceName == "" {
            options.ServiceName = DefaultServiceName
        }

        if options.Authenticator == nil {
-               options.Authenticator, err = core.GetAuthenticatorFromEnvironment(options.ServiceName)
+               options.Authenticator, err = auth.GetAuthenticatorFromEnvironment(options.ServiceName)
```

and the test for `GetAuthenticatorFromEnvironment` [here](https://github.com/IBM/cloudant-go-sdk/pull/15/files#diff-e3f5afd3c971deb73abb2594f3c44ba5d089afca6b094da568ad626a7a0a6fa9R31-R47), show that proper authenticator will be correctly selected.

This way the size of patch to the generated code will be minimal, there will be no need for additional test for  `NewCloudantV1UsingExternalConfig` or `NewCloudantV1` and other awkward workarounds.

In short, if I demonstrate that this authenticator handles base service's url and headers changes without all of these additional layers will that be work for you?

File: auth/session.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("Invalid format for AuthSession: %s", err)
+		}
+		expires = time.Unix(ts, 0)
+	}
+
+	return &session{
+		cookie:      c,
+		expires:     expires,
+		refreshTime: expires.Add(-(time.Until(expires) * 20 / 100)),
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/session.go:55]: This will not work, golang's `time` works with integers, but not with floats, it's hard to reason about time with  floating points, so it'll get truncated to 0.

To be honest I think this bit is pretty self-explanatory, but sure, I'll add the comment

File: auth/session_test.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		Expect(err).To(BeNil())
+		Expect(s).ToNot(BeNil())
+
+		// 20% of 24h is 4h48m
+		expected := 4*time.Hour + 48*time.Minute
+		roundedExpiration := s.expires.Sub(s.refreshTime).Round(time.Minute)
+		Expect(roundedExpiration).To(Equal(expected))
+	})
+
+	It("Test Session validation", func() {
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/session_test.go:78]: Tested below at line 94

File: auth/session_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		cookie := &http.Cookie{
+			Name:    "AuthSession",
+			Value:   makeAuthToken(),
+			Expires: time.Now().Add(time.Hour),
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/session_test.go:82]: It's generally better to not spread testing logic across different tests, but make unit test to test a unit of functionality. This test validated that `isValid` method works for both valid and possible invalid session states.

File: auth/authenticator_test.go (hunk 3 above)
Comment (eiri) [reply auth/authenticator_test.go:186]: This tests the code path where session's `needsRefresh()` called in `getCookie()` is `false`, the test below tests path where it's `true`

File: auth/authenticator_test.go (hunk 2 above)
Comment (eiri) [reply auth/authenticator_test.go:146]: Those two fetches here to show that cache works as cache when session's not expired, without it the test might give false positive in case cache fetches cookie every time it's called.

File: auth/authenticator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: In case of IAM you need to ask token from a Token Server `https://iam.cloud.ibm.com/identity/token`, so you need to open a new channel in the network towards this server.

But with CouchdbSessionAuthenticator you connect to Cloudant Server, towards an already opened direction in the network. For this server a Client is already defined, and maybe some connections already built up and those could be used for the authentication.
Also different transport settings could be used in the client I found here a good list about the settings: https://golang.org/pkg/net/http/#RoundTripper

And which option will be implemented from the listed 3 variations?

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go]: > But with CouchdbSessionAuthenticator you connect to Cloudant Server, towards an already opened direction in the network. 

Not really, this is authenticator, so it called first, before any connection opened and then it holds its own client with transport, that does all the things you mentioned. If one needs custom transport it can be set on authenticator's client directly, same way as for IAM, that's why `Client` var is public.

File: auth/couchdb_session_authenticator.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+// GetAuthenticatorFromEnvironment instantiates an Authenticator
+// using service properties retrieved from external config sources.
+func GetAuthenticatorFromEnvironment(credentialKey string) (authenticator core.Authenticator, err error) {
+	props, err := core.GetServiceProperties(credentialKey)
+	if err != nil {
+		return
+	}
+	if props[core.PROPNAME_AUTH_TYPE] == "COUCHDB_SESSION" {
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator.go:49]: The comparison is not case insensitive. We shall allow the `auth_type` is defined with lower case characters, too.

File: auth/authenticator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: What I also found, that the SSL verification settings is stored in the Client: https://github.com/IBM/go-sdk-core/blob/775201f679d69afb97f3259ababb24d482089410/v4/core/base_service.go#L184
What is the alternate solution to transfer this setting to the authenticator?

**Comment by vmatyusGitHub** [comment]
Unfortunately, my computer had problems during the meeting, so I could not really listen all what you said, tried to stop it from crashing continuously. I heard there is some limitation that you found during the embedding, but I could not understand fully what it was.

But okay, let's see your version with the URL and headers handling.

**Comment by vmatyusGitHub** [comment]
Please, note there is also context extension will be in the next version that this authenticator should be able to use the user specified context.

**Comment by vmatyusGitHub** [comment]
Okay, let away the `CouchdbBaseService` concept, I will accept the PR if the network related informations are transferred to the authenticator. I will collect them here:

- [ ] URL
- [ ] Headers
- [ ] context
- [ ] proxy
- [ ] DisableSSL

File: auth/authenticator_test.go (hunk 2 above)
Comment (vmatyusGitHub) [reply auth/authenticator_test.go]: Then I would like to request a comment above this line that this is the first cache check, and an update to the second attempt, that is the second request from the cache.

File: auth/session_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/session_test.go:82]: This test also can be interpreted as a negative test, because in other cases there was positive test. But okay.

File: auth/authenticator_test.go (hunk 2 above)
Comment (eiri) [reply auth/authenticator_test.go]: yeah, absolutely, it's evident it's too subtle and warrants the comment

File: auth/couchdb_session_authenticator.go (hunk 9 above)
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: Good catch! Addressed in https://github.com/IBM/cloudant-go-sdk/pull/15/commits/d610393a930b1f1abaf68e56a122db616b239a75

File: auth/session.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// needsRefresh atomically identifies if cookie is near of the expiration
+func (s *session) needsRefresh() bool {
+	now := time.Now()
+	if now.After(s.refreshTime) {
+		refreshMutex.Lock()
+		defer refreshMutex.Unlock()
+
+		// advance refresh time by one minute from now to avoid race
```
Comment (emlaver) [inline auth/session.go:79]: Maybe we can add a sentence (or two) about how this prevents threads waiting on mutex from starting the already called refresh process.

**Review by emlaver — CHANGES_REQUESTED** [review]

File: auth/authenticator.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		return err
+	}
+
+	request.AddCookie(cookie)
+	return nil
+}
+
+// getCookie returns an AuthSession cookie to be used in a request.
+// Whenever a new cookie is needed it fetched from the session end-point.
+func (a *CouchDbSessionAuthenticator) getCookie() (*http.Cookie, error) {
```
[the remark this answers is not in the document]
Comment (ricellis) [reply auth/authenticator.go]: There is no requirement _per se_. It's nice to follow the same patterns because it makes it easier for us to recollect across all the langs. It's also nice to use core code where possible because:
1. It gives us less of a maintenance problem (if, say, a fix is made in the core we're less likely to notice/consume that fix if we have duplicated).
1. The code gets more coverage/exposure when it is used by more SDKs & authenticator types so bugs are more likely to be found/flushed from it.

That said there is a balance to be struck between the complexity of trying to reuse core code vs the complexity of the code itself, which in reality differs for every language.

File: auth/authenticator_test.go (hunk 3 above)
Comment (vmatyusGitHub) [reply auth/authenticator_test.go]: Then would be good to add a comment here to note this part will get token from cache, because refresh time is not outdated.

File: auth/couchdb_session_authenticator.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	resp, err := a.Client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		buff := new(bytes.Buffer)
+		_, _ = buff.ReadFrom(resp.Body)
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go:201]: nah, we already in an error - if we can't read the original error message from a buffer nothing we can do about it and returning user an authentication error that reads "can't read an error from a buffer" is more confusing than just returning standard error without detailed description.

fwiw IAM authenticator does the same and I suspect for the same reasons

File: auth/session.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+// isValid checks if session has the cookie and it hasn't expired yet
+func (s *session) isValid() bool {
+	if s.cookie != nil && time.Now().Before(s.expires) {
+		return true
+	}
+	return false
+}
+
+// needsRefresh atomically identifies if cookie is near of the expiration
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/session.go:72]: No, atomically, as in "atomic operation". 😂

File: auth/couchdb_session_authenticator.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (emlaver) [reply auth/couchdb_session_authenticator.go:202]: Ok sounds good

File: auth/authenticator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: According the current implementation there is two client used, one for the server and an other one to the same server. There will be 2 network tunnels opened to the same address.

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go]: Yes, that's right, I thought we already established that. What about it?

[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: I cannot see where these transport options are transferred. What code should be created by the user?

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go]: @vmatyusGitHub I was busy with other stuff, give me some time to finish with url changes here that you requested and I'll show you the example, ok?

[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: Yeah, just take your time.

**Review by emlaver — APPROVED** [review]
Looks good

**Comment by mojito317** [comment]
One comment here.

Please update the README with session authentication. And do not forget to replace the `Basic` authentication recommendation for `Session` too.

**Review by vmatyusGitHub — COMMENTED** [review]
Now it looks much better here are my current findings, and also updated my collection comment, where I am following the network information that are transferred:
https://github.com/IBM/cloudant-go-sdk/pull/15#issuecomment-714569305

File: auth/couchdb_session_authenticator_test.go [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:169]: Here would be good to check the cookie content of the request is set properly with the `Authenticate` function.
and also in `Test authentication refresh`.

File: auth/couchdb_session_authenticator_test.go [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+			Expect(err).To(BeNil())
+
+			// Fetch a cookie from the cache to confirm it's present
+			// and prepare cache state for the next test
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:173]: What is preparation for the cache state?
```suggestion
			// Fetch a cookie from the session cache to confirm it's present
```

File: auth/couchdb_session_authenticator_test.go [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+			cookie, err := auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie).ToNot(BeNil())
+			Expect(cookie.Name).To(Equal("AuthSession"))
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(cookie.Expires).ToNot(BeNil())
+
+			// Fetch cookie again to verify cache is working as expected,
+			// we are getting old cookie and refresh process's not triggered
+			oldRefresh := auth.session.refreshTime
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:183]: 1. This check should be made with the first cache fetch, too.
2. I cannot see the point of this second cache fetch.

File: auth/couchdb_session_authenticator_test.go [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+			cookie, err = auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(auth.session.refreshTime).To(Equal(oldRefresh))
+
+			// Force expiration and verify we got a new cookie
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:189]:
```suggestion
			// Force expiration
```

File: auth/couchdb_session_authenticator_test.go [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+
+			// Run getCookie in three parallel threads, to verify
+			// that request mutex works and we are querying /_session only once
+			var wg sync.WaitGroup
+
+			for i := 1; i <= 3; i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					cookie, err = auth.getCookie()
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:200]: After this line you should add verify the test creates only 1 new cookie.

File: auth/couchdb_session_authenticator_test.go [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+					Expect(cookie.Value).To(Equal("fakefake-2"))
+					Expect(auth.session.refreshTime).ToNot(Equal(oldRefresh))
+				}()
+			}
+			wg.Wait()
+			close(done)
+		})
+
+		It("Test authentication refresh", func(done Done) {
+			err = auth.Authenticate(request)
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:211]: Check that the cookie is attached to the `request` object.

File: auth/session.go [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+	// refreshTime is 20% of period between now and the expiration time
+	return &session{
+		cookie:      c,
+		expires:     expires,
+		refreshTime: expires.Add(-(time.Until(expires) * 20 / 100)),
+	}, nil
+}
+
+func (s *session) getCookie() *http.Cookie {
```
Comment (vmatyusGitHub) [inline auth/session.go:60]: Can you remove this `getCookie` function and use the session referencing instead like `s.cookie`?

**Review by vmatyusGitHub — COMMENTED** [review]
Findings during checking the comment corrections.

File: auth/authenticator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmatyusGitHub) [reply auth/authenticator.go]: I found an example in t your gist: https://gist.github.com/eiri/f9e6a2a7ed7e0cc6bdf7a8fec76c387b#just-use-public-field-client
Did you noticed there is a CookieJar in the Client: https://golang.org/pkg/net/http/#Client ?

[the remark this answers is not in the document]
Comment (eiri) [reply auth/authenticator.go]: Yes, I'm familiar with standard golang's packages

File: auth/couchdb_session_authenticator_test.go (hunk 14 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:169]: This checked on the line right below this one in the block of test around `auth.getCookie()` and same in `Test authentication refresh`

File: auth/couchdb_session_authenticator_test.go (hunk 15 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:173]: Cache state includes cookie, its expiration time and its refresh time. In this test we are not as much interested in presence of the cookie, but in times around its management.

File: auth/couchdb_session_authenticator_test.go (hunk 17 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:189]: This describes not the single line below it, but the whole the rest of the block

File: auth/couchdb_session_authenticator_test.go (hunk 18 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:200]: I don't understand what you are saying here, this is exactly what next three test are verifying. Please elaborate.

File: auth/session.go (hunk 20 above)
Comment (eiri) [reply auth/session.go:60]: No, that'd be bad design with leaking internals of `session` struct.

File: auth/couchdb_session_authenticator_test.go (hunk 19 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:211]: Why? We are not testing `http.Request` and its `AddCookie` function, we are testing authenticator's refresh correctness, why suddenly break tests isolaton here?

File: auth/couchdb_session_authenticator_test.go (hunk 16 above)
Comment (eiri) [reply auth/couchdb_session_authenticator_test.go:183]: 1. What check? I'm storing old refresh time here so further on I can confirm the refresh time will advance.
1. Right, I see you are struggling with this test, consider number of the comments you left. I goes like this:

- Do `auth.Authenticate` to get cookie and its times into the cache.
- Do first cache fetch and confirm `Authenticate` worked and we have out cookie and its times.
- Store refresh time, do _second_ cache fetch and verify we still have old cookie and refresh time, to confirm that we didn't accidentally triggered offline refresh with the _previous_ cache fetch.
- Move expiration time, do three cache fetches in parallel and verify that we've got one and the same new cookie and new refresh time, so our mutex works and we only starting one offline cache refresh.

In other words second fetch here is to verify that we are not accidentally triggering offline cache refresh on every fetch. Hope this helps.

File: auth/couchdb_session_authenticator.go [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+const (
+	AUTHTYPE_COUCHDB_SESSION = "COUCHDB_SESSION"
+)
+
+var requestSessionMutex sync.Mutex
+
+// CouchDbSessionAuthenticator uses username and password to obtain
+// CouchDB authentication cookie, and adds the cookie to requests.
+type CouchDbSessionAuthenticator struct {
+	Username, Password string
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go:41]: In 4ce5410

File: auth/couchdb_session_authenticator.go [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+	Client             *http.Client
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: In 4ce5410

File: auth/couchdb_session_authenticator.go [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+			}
+		}
+		return authenticator, err
+	}
+
+	return core.GetAuthenticatorFromEnvironment(credentialKey)
+}
+
+// NewCouchDbSessionAuthenticator constructs a new NewCouchDbSessionAuthenticator instance.
+func NewCouchDbSessionAuthenticator(username, password string) (*CouchDbSessionAuthenticator, error) {
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go:64]: 😒   in 4ce5410 if this makes you happy

File: auth/couchdb_session_authenticator.go [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+	if a.session != nil && a.session.isValid() {
+		return nil
+	}
+
+	err := a.requestSession()
+	return err
+}
+
+// requestSession fetches new AuthSession cookie from session end-point.
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: Is it really an improvement? To me this sounds exactly the same, but whatever. In 4ce5410

File: auth/couchdb_session_authenticator.go [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+
+// CouchDbSessionAuthenticator uses username and password to obtain
+// CouchDB authentication cookie, and adds the cookie to requests.
+type CouchDbSessionAuthenticator struct {
+	Username, Password string
+	Client             *http.Client
+	url                string
+	header             http.Header
+	ctx                context.Context
+	disableSSL         bool
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: This variable is not intended to be programmatically mutable, unlike IAM this is served by the same server, so it less error prune to inherit it from config.

Renamed to longer name in 4ce5410

**Review by vmatyusGitHub — APPROVED** [review]
Considering the need of this feature +1

**Comment by eiri** [comment]
Cool beans. @emlaver @vmatyusGitHub - thank you for the reviews! I'm bringing this in now.
